IRremoteESP8266.cpp

```diff
@@ -427,24 +427,76 @@ void ICACHE_FLASH_ATTR IRsend::sendRCMM(uint32_t data, uint8_t nbits) {
 
 }
 
+// Send a Panasonic formatted message.
+//
+// Args:
+//   data:   The message to be sent.
+//   nbits:  The number of bits of the message to be sent. (PANASONIC_BITS).
+//   repeat: The number of times the command is to be repeated.
+//
+// Note:
+//   This protocol is a modified version of Kaseikyo.
+void ICACHE_FLASH_ATTR IRsend::sendPanasonic64(unsigned long long data,
+                                               unsigned int nbits,
+                                               unsigned int repeat) {
+  enableIROut(37);  // Set IR carrier frequency
+  IRtimer usecTimer = IRtimer();
+
+  for (uint16_t i = 0; i <= repeat; i++) {
+    usecTimer.reset();
+    // Header
+    mark(PANASONIC_HDR_MARK);
+    space(PANASONIC_HDR_SPACE);
+    // Data
+    sendData(PANASONIC_BIT_MARK, PANASONIC_ONE_SPACE,
+             PANASONIC_BIT_MARK, PANASONIC_ZERO_SPACE,
+             data, nbits, true);
+    // Footer
+    mark(PANASONIC_BIT_MARK);
+    space(max(PANASONIC_MIN_COMMAND_LENGTH - usecTimer.elapsed(),
+              PANASONIC_MIN_GAP));
+  }
+}
+
+// Send a Panasonic formatted message.
+//
+// Args:
+//   address: The manufacturer code.
+//   data:    The data portion to be sent.
+//   nbits:   The number of bits of the message to be sent. (PANASONIC_BITS).
+//   repeat:  The number of times the command is to be repeated.
+//
+// Note:
+//   This protocol is a modified version of Kaseikyo.
 void ICACHE_FLASH_ATTR IRsend::sendPanasonic(unsigned int address,
-                                             unsigned long data) {
-  // Set IR carrier frequency
-  enableIROut(35);
-  // Header
-  mark(PANASONIC_HDR_MARK);
-  space(PANASONIC_HDR_SPACE);
-  // Address (16 bits)
-  sendData(PANASONIC_BIT_MARK, PANASONIC_ONE_SPACE,
-           PANASONIC_BIT_MARK, PANASONIC_ZERO_SPACE,
-           address, 16, true);
-  // Data (32 bits)
-  sendData(PANASONIC_BIT_MARK, PANASONIC_ONE_SPACE,
-           PANASONIC_BIT_MARK, PANASONIC_ZERO_SPACE,
-           data, 32, true);
-  // Footer
-  mark(PANASONIC_BIT_MARK);
-  ledOff();
+                                             unsigned long data,
+                                             unsigned int nbits,
+                                             unsigned int repeat) {
+  sendPanasonic64(((uint64_t) address << 32) | (uint64_t) data, nbits, repeat);
+}
+
+// Calculate the raw Panasonic data based on device, subdevice, & function.
+//
+// Args:
+//   device:    An 8-bit code.
+//   subdevice: An 8-bit code.
+//   function:  An 8-bit code.
+// Returns:
+//   A raw uint64_t Panasonic message.
+//
+// Note:
+//   Panasonic 48-bit protocol is a modified version of Kaseikyo.
+// Ref:
+//   http://www.remotecentral.com/cgi-bin/mboard/rc-pronto/thread.cgi?2615
+unsigned long long ICACHE_FLASH_ATTR IRsend::encodePanasonic(uint8_t device,
```

Comment:
It's a public function. We don't use it ourselves in the library, but end-users might find it useful to construct IR codes for Panasonic devices where they don't have _our_ version of codes or the ability to capture them. e.g. Some sites list the sub-codes, rather than 48bit blob (or GC format).